core/arch/arm/tee/pta_generic.c

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Intel Corporation
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <kernel/msg_param.h>
+#include <optee_rpc_cmd.h>
+#include <pta_generic.h>
+#include <string.h>
+#include <tee/tee_fs_rpc.h>
+
+static uint32_t get_instance_id(struct tee_ta_session *sess)
+{
+	return sess->ctx->ops->get_instance_id(sess->ctx);
+}
+
+static inline bool is_param_memref(uint32_t param_types, uint32_t idx)
+{
+	uint32_t ptype = TEE_PARAM_TYPE_GET(param_types, idx);
+
+	if ((ptype == TEE_PARAM_TYPE_MEMREF_INPUT) ||
+			(ptype == TEE_PARAM_TYPE_MEMREF_OUTPUT) ||
+			(ptype == TEE_PARAM_TYPE_MEMREF_INOUT))
+		return true;
+	return false;
+}
+
+static inline bool is_param_out(uint32_t param_types, uint32_t idx)
+{
+	uint32_t ptype = TEE_PARAM_TYPE_GET(param_types, idx);
+
+	if ((ptype == TEE_PARAM_TYPE_VALUE_OUTPUT) ||
+			(ptype == TEE_PARAM_TYPE_VALUE_INOUT) ||
+			(ptype == TEE_PARAM_TYPE_MEMREF_OUTPUT) ||
+			(ptype == TEE_PARAM_TYPE_MEMREF_INOUT))
+		return true;
+	return false;
+}
+
+static inline bool is_param_value(uint32_t param_types, uint32_t idx)
+{
+	uint32_t ptype = TEE_PARAM_TYPE_GET(param_types, idx);
+	if (ptype == TEE_PARAM_TYPE_VALUE_INPUT ||
+			ptype == TEE_PARAM_TYPE_VALUE_OUTPUT ||
+			ptype == TEE_PARAM_TYPE_VALUE_INOUT)
+		return true;
+
+	return false;
+}
+
+static inline bool is_param_none(uint32_t param_types, uint32_t idx)
+{
+	if (TEE_PARAM_TYPE_GET(param_types, idx) == TEE_PARAM_TYPE_NONE)
+		return true;
+	return false;
+}
+
+static void *alloc_transient_shm(size_t size, struct mobj **mobj)
+{
+	paddr_t p;
+	void *va;
+
+	*mobj = thread_rpc_alloc_payload(size);
+	if (!*mobj)
+		return NULL;
+
+	if (mobj_get_pa(*mobj, 0, 0, &p))
+		goto err;
+
+	if (!ALIGNMENT_IS_OK(p, uint64_t))
+		goto err;
+
+	va = mobj_get_va(*mobj, 0);
+	if (!va)
+		goto err;
+
+	return va;
+
+err:
+	thread_rpc_free_payload(*mobj);
+	return NULL;
+}
+
+static void free_transient_shm(struct mobj *mobj)
+{
+	thread_rpc_free_payload(mobj);
+}
+
+static void *prepare_memref_params(TEE_Param *param, uint32_t param_type,
+			 bool cached, struct mobj **mobj,
+			 struct thread_param *tpm)
+{
+	void *va = NULL;
+	size_t size = param->memref.size;
+
+	if (cached)
```

Comment:
How about letting the TA pre-allocate shared memory buffers via another new TA function instead? It would simplify this PTA and also enable zero copy.

Reply:
Hmm. I think it will complicate the calling mechanism. Currently, it is similar to calling another UTA (internally it calls PTA). 

Anyways, it's interesting to have that feature, may be using another PTA, called a zero-copy REE Service. It will also take some good time to implement, so, may be as a next step please.

Reply:
Yes, and that mess (when a UTA is calling another UTA and temporary buffers must be allocated inside OP-TEE Core) is something i regret now since it's hard to get rid of the code while maintaining backwards compatibility.

I can add a function to allocate non-secure shared memory from a TA.

Reply:
Currently, from normal world, we can create a Shared memory using Global Platform APIs. I have not explored that interface for now, but it could be another way. However, the programming model will involve interleaving the calls to REE service with buffer fill. It's just that as a first step, it will be a bit of complicated flow to present. Also, I hope that users  are careful enough not to start using the shared buffer for TA internal buffers, thinking that I am not exhausting secure RAM ;).